src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs

```diff
@@ -0,0 +1,185 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers.Binary;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+
+namespace System.Reflection.Emit
+{
+    internal readonly struct CustomAttributeWrapper
+    {
+        private readonly ConstructorInfo _constructorInfo;
+        private readonly byte[] _binaryAttribute;
+
+        public CustomAttributeWrapper(ConstructorInfo constructorInfo, ReadOnlySpan<byte> binaryAttribute)
+        {
+            _constructorInfo = constructorInfo;
+            _binaryAttribute = binaryAttribute.ToArray(); // TODO: Update to BlobHandle when public API public APi for MetadataBuilder.GetOrAddBlob(ReadOnlySpan<byte>) added
+        }
+
+        public ConstructorInfo Ctor => _constructorInfo;
+        public byte[] Data => _binaryAttribute;
+    }
+
+    internal struct CustomAttributeInfo
+    {
+        public ConstructorInfo _ctor;
+        public object?[] _ctorArgs;
+        public string[] _namedParamNames;
+        public object?[] _namedParamValues;
+
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2057:Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'",
```

Comment:
What makes it safe to suppress this warning?

Reply:
The warning come up when loading named parameters info from Ctor:
- It warns that we could not use `[DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.NonPublicFields | DynamicallyAccessedMemberTypes.PublicFields)]` attribute for ConstructorInfo
- For `IL2057:Unrecognized value passed to the parameter 'typeName' of method 'System.Type.GetType(String)'` I thought its impossible to handle it because the `typeName` loaded at runtime. Now I think we could propogate the `[RequiresUnreferencedCode]` attribute, though that would cause the attribute to be propagated up to the public API `public void SetCustomAttribute(ConstructorInfo con, byte[] binaryAttribute)`, not sure we would like to do that ...

Reply:
Is the `byte[]` path publicly accessible or do we construct the `byte[]` from user-provided `PropertyInfo`s/`FieldInfo`s? If we can guarantee there was a `PropertyInfo`/`FieldInfo` for these before and we're just roundtripping it through byte array, this is safe and could be suppressed (however we have to be very careful not to break the invariant in the future).

But if user can provide arbitrary `byte[]`, this cannot be suppressed and needs to be bubbled up to the public API with the `byte[]`. Alternatively, we could require the user to provide both the `Type` (with the proper annotation) and a `ConstructorInfo`. This is obviously a bit awkward.

Reply:
`SetCustomAttribute(ConstructorInfo con, byte[] binaryAttribute)` is a public API and we cannot guarantee that `PropertyInfo/FieldInfo` that needed for parsing the byte[] kept in `ConstructorInfo.DeclaringType`. 

Because it is a public API `requiring the user to provide both the Type (with the proper annotation) and a ConstructorInfo` is not an option. 

For now I have only option to apply [RequiresUnreferencedCode] attribute with better messaging, something like: `[RequiresUnreferencedCode("The named parameters used for the attribute might be trimmed")]`

Reply:
The problem is that the custom attribute parser is not appropriate for what we need to do here. The custom algorithm parsing should not be creating instances of types referenced by the custom attribute blob.

Reply:
Good point, done, thanks!